[fix/keyboards-shortcuts] Added missing keyboard shortcuts / fix broken keyboard navigation

hosy opened the pull request:
## Description

- added missing keyboard shortcuts in PDF detail view (Go to Page, Search)
- added keyboard shortcuts for media playback
- added missing more action keyboard shortcuts in detail view
- fixed broken keyboard shortcuts arrow navigation in UITableView between sections

## Related Issue
#902 
#903

## Motivation and Context
Access all functions via keyboard shortcuts

## How Has This Been Tested?
Check issues

## Screenshots (if appropriate):

![Simulator Screen Shot - iPad Pro (11-inch) (2nd generation) - 2021-02-22 at 19 53 21](https://user-images.githubusercontent.com/736109/108860196-437f1400-75ee-11eb-9d2d-87c88d325852.png)


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Added changelog files for the fixed issues in folder changelog/unreleased

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.6/New%20keyboard%20shortcuts.md

- [X] (1) Shortcuts to navigate and close the search view in PDF https://github.com/owncloud/ios-app/pull/908#issuecomment-792848448
- [x] (2) Open in https://github.com/owncloud/ios-app/pull/908#issuecomment-792857377
- [x] (3) Rewind and fast forward https://github.com/owncloud/ios-app/pull/908#issuecomment-792864260
- [x] (4) Missing shortcuts https://github.com/owncloud/ios-app/pull/908#issuecomment-792877796
- [x] (5) Jump to the fist item on the list https://github.com/owncloud/ios-app/pull/908#issuecomment-792886062
- [x] (6) Copy/move picker not allow to select items with the keyboardhttps://github.com/owncloud/ios-app/pull/908#issuecomment-793702855
- [X] (7) Links missing in PDF preview https://github.com/owncloud/ios-app/pull/908#issuecomment-793711770

mneuwert commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -274,6 +275,9 @@ extension UITableViewController {
 	@objc func selectNext(sender: UIKeyCommand) {
 		if let selectedIndexPath = self.tableView?.indexPathForSelectedRow, selectedIndexPath.row < (self.tableView?.numberOfRows(inSection: selectedIndexPath.section) ?? 0 ) - 1 {
 			self.tableView.selectRow(at: NSIndexPath(row: selectedIndexPath.row + 1, section: selectedIndexPath.section) as IndexPath, animated: true, scrollPosition: .middle)
+		} else if let selectedIndexPath = self.tableView?.indexPathForSelectedRow, (selectedIndexPath.section + 1) < self.tableView?.numberOfSections ?? 0 {
```
@hosy for readability I would suggest to add some guard let statements in the beginning for all the optional stuff

hosy replied:
Fixed!

mneuwert commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -282,6 +286,11 @@ extension UITableViewController {
 	@objc func selectPrevious(sender: UIKeyCommand) {
 		if let selectedIndexPath = self.tableView?.indexPathForSelectedRow, selectedIndexPath.row > 0 {
 			self.tableView.selectRow(at: NSIndexPath(row: selectedIndexPath.row - 1, section: selectedIndexPath.section) as IndexPath, animated: true, scrollPosition: .middle)
+		} else if let selectedIndexPath = self.tableView?.indexPathForSelectedRow, selectedIndexPath.row == 0, selectedIndexPath.section > 0 {
```
@hosy again some guard lets upfront would make this code easier to read I think...

hosy replied:
Fixed!

mneuwert commented on ownCloud/Client/Viewer/Media/MediaDisplayViewController.swift:
```diff
@@ -32,7 +32,7 @@ class MediaDisplayViewController : DisplayViewController {
 	private var playerItemStatusObservation: NSKeyValueObservation?
 	private var playerItem: AVPlayerItem?
-	private var player: AVPlayer?
+	var player: AVPlayer?
```
@hosy eventually don't expose the player but may be add some properties you need accessing `AVPlayer`? As an example I've seen using `AVPlayer.rate` for checking if it is playing something? Instead you could add `MediaDisplayViewController.isPlaying`

hosy replied:
Fixed!

mneuwert commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -1016,13 +1042,90 @@ extension DisplayHostViewController {
 	}
 
+	@objc func performMoreItemAction(_ command : UIKeyCommand) {
+		if let core = core, let currentViewController = self.viewControllers?.first as? DisplayViewController, let item = currentViewController.item {
```
@hosy try avoiding long lines for readability

hosy replied:
Fixed!

mneuwert commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -1016,13 +1042,90 @@ extension DisplayHostViewController {
 			shortcuts.append(previousObjectCommand)
 		}
 
+		if let core = core, let currentViewController = self.viewControllers?.first as? DisplayViewController, let item = currentViewController.item {
```
@hosy try avoiding long lines for readability

hosy replied:
Fixed!

hosy commented:
@mneuwert code review findings are fixed and you can give it a new try. Thanks!

jesmrec commented:
### (1)

1. Open PDF
2. keyboard: cmd + L  (search)

Current: Results after searching can not be navigated. There is no keyboard shortcut to close the search menu.
Expected: Shortcuts to navigate and to close the view. 

iPadAir iOS13
`389cd13`

jesmrec commented:
### (2)

1. Open a file to preview
2. keyboard: cmd + O  (open in)


Current: Not visible:

![Screen Shot 2021-03-08 at 14 52 59](https://user-images.githubusercontent.com/14894746/110346166-936acb80-802f-11eb-945d-ce9041447438.png)

Expected: visible

iPadAir iOS13
`389cd13`

NOTE: probably a regression. It is up to you to fix here or address it in a separate issue.

jesmrec commented:
### (3)

In the media player, there are two shortcuts:

Rewind:   ^ + ←
Fast Forward:  ^ +  →

I could not use them. When i click the keys it jumps to the next/previous media file. (as the proper shortcut with that purpose)

iPadAir iOS13
`389cd13`

jesmrec commented:
### (4) (improvement)

This is an improvement, so it could be addressed in a separate ticket.

In the media player i miss two shortcuts:

- For the mute/unmute button (top right corner)
- For the full screen/windowed option (top left corner)

jesmrec commented:
### (5)

you mention:

`fixed broken keyboard shortcuts arrow navigation in UITableView between sections`

not sure if it has something to do with the following scenario:

1. using the keyboard, add a new bookmark
2. Browse into it ( keyboard:  → )

Current: In the files view, i did not find any (straightforward) way to select items on the list. I found a shortcut with the "Select last item on the page" (cmd + down), but not a way to select, f.ex. , the first item on the list.

Is that the expected?

It was tested with iOS13, does it affect?

Probably not a regression.

iPadAir iOS13
`389cd13`

jesmrec commented:
### (6)

Probably related with (5). Same problem happens after opening the move/copy picker. I can open it, but no way to select an item inside.

jesmrec commented:
### (7)

1. Open PDF file
2. Long press on cmd to get the list on available commands

Current: No shortcut to `Links` 
![Screen Shot 2021-03-09 at 11 52 01](https://user-images.githubusercontent.com/14894746/110460155-1e9a9e80-80ce-11eb-93e4-9f78231b24f6.png)

Expected: Shortcut to links

NOTE: this one is not a bug, and could be addressed in a separate issue 

iPadAir iOS13
`389cd13`

hosy commented:
@jesmrec 
  (2) Open in
fixed

  (3) Rewind and fast forward
for me it is working. Are you sure you are using `Control` + `Arrow left/right`. It sounds you are using `Command` + `Arrow left/right`

  (5) Jump to the fist item on the list
  (6) Copy/move picker not allow to select items with the keyboard
both are fixed now (regression after a CR fix)

  (7) Links missing in PDF preview
fixed

I will have a look at (1) and (4) now and will check how many effort is needed.

hosy commented:
@jesmrec (4) 
- added keyboard commands for fullscreen and mute/unmute 
- changed some shortcuts to system wide used
- changed some labels for shortcuts to system wide used

hosy commented:
@jesmrec (1) 
Added keyboard shortcuts for navigating in the PDF search results table view

Now every finding should be addressed now!

jesmrec commented:
(1) -> now it is posible to navigate through list of results. I'd add a new hint to the following existing ones: the right arrow to submit the selected item:

![Screenshot 2021-03-19 at 09 46 19](https://user-images.githubusercontent.com/14894746/111754189-fd893900-8897-11eb-8bca-96bca39ec138.png)

OTOH, there is no way to navigate through multiple results in the pdf. Should we address in new issue or fix here?

![Screenshot 2021-03-19 at 09 47 47](https://user-images.githubusercontent.com/14894746/111754355-2e696e00-8898-11eb-8830-9c75ab082fa2.png)


(2) -> fixed

(3) -> probably you were right and i was probably using cmd instead of ctrl. It works fine!

(4) -> done

(5) -> fixed

(6) -> fixed

(7) -> new shortcuts for links/share are not easy. F .ex. if i use `^ + cmd + L`  to open links, most of the times is opened the  Search menu (`cmd + L`). I'd combine alt or ctrl with the keys instead of ^ (this is a very personal opinion, though). What do you think?

hosy commented:
@jesmrec
regarding (1): If you select an item in the search results, the command is available to select it
In the latest commit I added keyboard shortcuts to select the next and previous search result and to close the search results.

(7) I changed the keyboard command for `Public Links` to `CMD + L` and the `PDF Search` to `CMD + S`

Furthermore in the directory picker the keyboard commands are only available, if an item is selected and the item is a folder (for the select command)

jesmrec commented:
(1) and (7) fixed.

This is approved on my side.
